### PR TITLE
[ENH] Added batchTC and added option to remove mutual info from factor-VAE

### PR DESCRIPTION
**Pull request by linesd**
- Added batchTC from the paper Isolating Sources of Disentanglement in VAE's
- Added option to remove the mutual information term from the loss calculation in Factor-VAE.

Note: There is a quick fix at the end of BatchTCLoss regarding the visualisation (I have't had time to dive into the vis stuff and fix the issue - Yann, your suggestion of adding an extra argument to the function was still causing issues). Currently I am storing some tc_loss values as a substitute for the kl_loss of each dimension to get around the indexing issue for visualisation. This will need to be removed once the viz is fixed.

### CODE REVIEW

File: disvae/losses.py [hunk 1]
```diff
@@ -240,6 +264,170 @@ def __call__(self, data, model, optimizer, storer):
...
+            storer['tc_loss'].append(tc_loss.item())
+
+            # TODO Remove this when visualisation fixed
+            tc_loss_vec = (logqz - logqz_prodmarginals)
+            for i in range(self.latent_dim):
+                storer['kl_loss_' + str(i)].append(tc_loss_vec[i].item())
+
+        return loss
+
+    def _log_density_normal(self, latent_sample, latent_dist, batch_size, return_matrix=False):
```
Comment (YannDubs) [inline disvae/losses.py]: This should be in `utils` because it's not loss specific (I actually have such such function in the `eval.py`which was written during the 2 days poster sprint so not as clean. We don't want 2 functions that do the same in the codebase) either put it in `utils/helpers.py` or create a new module `utils/math.py`, but it's definitely more general than just being a private method. Of course if it's in `utils` it should not have `_*` . The same might be hold for `_log_importance_weight_matrix`, but I'm not sure what this one does yet hah

**Comment by linesd** [comment]
> A partial review of things that just stood up while scrolling: I'll do a full one next week.
> 
> Also, you seem to not have solved the viz issue. Did you try what I sent you on slack last time ? Because now it means you are computing `kl_normal_loss` when you don't need it, and you are storing something which is not what we want (when commenting `# TODO Remove this when visualisation fixed`)

Yeah I did try it and it was still causing issues. I'm not doing any extra calculations here, just storing info that we don't need to avoid the indexing issue. I realise its not a clean way to do this but I just don't have the time right now to try and fix the viz stuff and thought it's better that this code is up so we can run some experiments (as far as I know, this quick fix shouldn't affect our testing?).

File: disvae/losses.py [hunk 2]
```diff
@@ -165,26 +175,32 @@ class FactorKLoss(BaseLoss):
...
-        discriminator : disvae.discriminator.Discriminator
-
-        optimizer_d : torch.optim
-
         device : torch.device
 
         beta : float, optional
             Weight of the TC loss term. `gamma` in the paper.
 
+        mutual_info : bool
```
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/losses.py]: Yeah that's a nice convention, happy to go with that.

File: disvae/losses.py [hunk 3]
```diff
@@ -205,16 +221,24 @@ def __call__(self, data, model, optimizer, storer):
         rec_loss = _reconstruction_loss(data1, recon_batch, storer=storer)
         kl_loss = _kl_normal_loss(*latent_dist, storer)
         d_z = self.discriminator(latent_sample1)
+        
         # clamping to 0 because TC cannot be negative : TESTTTTTTTT
         tc_loss = (F.logsigmoid(d_z) - F.logsigmoid(1 - d_z)).clamp(0).mean()
-        vae_loss = rec_loss + kl_loss + self.beta * tc_loss
+
+        if self.mutual_info:
```
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/losses.py]: Happy to stick to your personal preferences here for code consistency. Will change the beta issue too.

**Comment by YannDubs** [comment]
> 
> Yeah I did try it and it was still causing issues. I'm not doing any extra calculations here, just storing info that we don't need to avoid the indexing issue. I realise its not a clean way to do this but I just don't have the time right now to try and fix the viz stuff and thought it's better that this code is up so we can run some experiments (as far as I know, this quick fix shouldn't affect our testing?).

Yeah no worries if it was not working, I thought it would be a simple copy / paste /solved issue. Then nor worries for storing additional things. But I'm still pretty sure you are computing additional things. YOu are computing `_kl_normal_loss` even when we don't use mutual information, but we don't need that (the problem is that if you don't you will have the `.gif` issue again). If you follow the code I sent above you will not anymore, so you will have to store dim_kl_loss for each dimension to solve that problem. ALternatively keep things like they are now (besides the beta which is important for testing), and just add a `#TO-DO` linking to this PR and saying that we should not be computing  `_kl_normal_loss` when no mutal information is used (might be quicker for you :))

File: disvae/losses.py [hunk 4]
```diff
@@ -205,16 +221,24 @@ def __call__(self, data, model, optimizer, storer):
         rec_loss = _reconstruction_loss(data1, recon_batch, storer=storer)
         kl_loss = _kl_normal_loss(*latent_dist, storer)
```
Comment (YannDubs) [inline disvae/losses.py]: this is where there's computation that is not used in case of `not is_mutual_information` just add a comment

**Comment by linesd** [comment]
> > Yeah I did try it and it was still causing issues. I'm not doing any extra calculations here, just storing info that we don't need to avoid the indexing issue. I realise its not a clean way to do this but I just don't have the time right now to try and fix the viz stuff and thought it's better that this code is up so we can run some experiments (as far as I know, this quick fix shouldn't affect our testing?).
> 
> Yeah no worries if it was not working, I thought it would be a simple copy / paste /solved issue. Then nor worries for storing additional things. But I'm still pretty sure you are computing additional things. YOu are computing `_kl_normal_loss` even when we don't use mutual information, but we don't need that (the problem is that if you don't you will have the `.gif` issue again). If you follow the code I sent above you will not anymore, so you will have to store dim_kl_loss for each dimension to solve that problem. ALternatively keep things like they are now (besides the beta which is important for testing), and just add a `#TO-DO` linking to this PR and saying that we should not be computing `_kl_normal_loss` when no mutal information is used (might be quicker for you :))

Ah yes, sorry. I was specifically talking about the batchTC loss. But you are correct, the _kl_normal_loss is unnecessarily computed in factor-vae loss. OK I'll leave it there for now with a TODO and a link to this convo.

File: disvae/losses.py [hunk 5]
```diff
@@ -240,6 +264,170 @@ def __call__(self, data, model, optimizer, storer):
...
+        mss : bool
+            Selects either minibatch stratified sampling (True) or minibatch
+            weighted  sampling (False)
+
+        References :
+           [1] Chen, Tian Qi, et al. "Isolating sources of disentanglement in variational
+           autoencoders." Advances in Neural Information Processing Systems. 2018.
+    """
+
+    def __init__(self, alpha, beta, gamma, data_size, latent_dim, mss=False):
```
Comment (YannDubs) [inline disvae/losses.py]: `is_mss` for consistency

File: disvae/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/losses.py]: Have a left TODO for this per our conversation regarding the visualisation issue.

File: disvae/losses.py (hunk 4 above)
Comment (linesd) [reply disvae/losses.py]: Added a TODO to remove once viz is fixed.

File: disvae/losses.py (hunk 5 above)
Comment (linesd) [reply disvae/losses.py]: Changed

File: disvae/losses.py (hunk 1 above)
Comment (linesd) [reply disvae/losses.py]: If ```_log_density_normal``` goes into a math.py then I think ```_log_importance_weight_matrix``` should as well. I'll sort this out.

File: disvae/losses.py (hunk 5 above)
Comment (YannDubs) [inline disvae/losses.py]: `is_mss` for consistency. Default hyperparameters for `alpha`, `beta`, `gamma` should be given. 

I don't see any place you actually need `latent_dim` better drop it for having similar constructor to the other losses. Everywhere I see `latent_dim` you can actually get it from the shapes of the input vectors (seem like it, I might be wrong)

Comment (linesd) [reply disvae/losses.py]: I have changed both of these.

**Comment by linesd** [comment]
> don't forget `(self.beta+1)*tc_loss` when no mutual info

I very almost left this commented out. Thanks :)

**Comment by linesd** [comment]
@alecokas  I have left a comment and link for you in the viz/visualize.py file regarding an issue with the gif generation. We are no longer storing all the KL values for all latent variable dimensions for the new loss I've added and this is causing issues.
